### ISSUE TITLE
Deprecate BingMaps (ideally replace with AzureMaps)

### ISSUE THREAD

**Issue by ChrisHSandN** [openlayers/openlayers#15853]
**Is your feature request related to a problem? Please describe.**
Today I received the following email from Bing Maps:

> On May 21, 2024, we shared our vision for a single enterprise mapping platform at Microsoft and that we are unifying our Bing Maps for Enterprise offering under Azure Maps. This means that going forward, Azure Maps will combine the best of Bing Maps for Enterprise and Azure Maps. Today we are announcing that as part of these unification efforts, Bing Maps for Enterprise will be retired. The ability to purchase Bing Maps for Enterprise SKUs and the availability of the Bing Maps for Enterprise services will end according to the following timeline:
> -    Customers using Bing Maps for Enterprise Basic or Free license will no longer be able to use Bing Maps for Enterprise services beyond June 30, 2025.
(a) Basic / Free Licenses customers who would like to continue using enterprise maps from Microsoft are encouraged to move to Azure Maps before June 30, 2025.
(b) Existing Basic / Free Licenses customers who would like to continue using Bing Maps for Enterprise until the service end date of June 30, 2028, can upgrade to a paid Enterprise License by contacting [maplicq@microsoft.com](mailto:maplicq@microsoft.com) before June 30, 2025.
> -    Bing Maps for Enterprise services will retire, and support will end on June 30, 2028, and customers will no longer be able to use any of the Bing Maps for Enterprise services.

**Describe the solution you'd like**

In light of this I was wondering if it would be worth deprecating the BingMaps module so future projects don't build projects against an API which is shutting down?

Azure maps published their own OL module: https://github.com/azure-samples/azure-maps-openlayers/tree/main/ so ideally it would be nice if this could be moved into the core of OL to replace the BingMaps module?  The Microsoft module hasn't been updated for years and doesn't work with OL5+ without webpack fudges https://github.com/Azure-Samples/azure-maps-openlayers/issues/2.

**Comment by ahocevar** [comment]
We're going to do that eventually anyway. If someone considers this more urgent, we're happy to accept a pull request.

**Comment by elliotlarson** [comment]
The plugin's code looked fairly old in a lot of places and overcomplicated.  Here's what I ended up doing in case this helps anyone else:

I created a factory for returning either the Bing or Azure tile layer `tile-layer-factory.js`:

```javascript
// Accepts a provider ("bing" or "azure"), and an API key (Bing API key or Azure maps primary key)

import TileLayer from "ol/layer/Tile";
import BingMaps from "ol/source/BingMaps";
import XYZ from "ol/source/XYZ";

export default class TileLayerFactory {
  constructor(provider, providerApiKey) {
    this.provider = provider;
    this.providerApiKey = providerApiKey;
  }

  get tileLayer() {
    return new TileLayer({
      visible: true,
      preload: Infinity,
      source: this.#source,
    });
  }

  get #source() {
    switch (this.provider) {
      case "bing":
        return this.#bingSource;
      case "azure":
        return this.#azureSource;
      default:
        new Error(`Unknown tile provider: ${this.provider}`);
    }
  }

  get #bingSource() {
    return new BingMaps({ key: this.providerApiKey, imagerySet: "AerialWithLabelsOnDemand" });
  }

  get #azureSource() {
    const url = "https://atlas.microsoft.com/map/tile";
    const params = [
      `subscription-key=${this.providerApiKey}`,
      "api-version=2.0",
      "tilesetId=microsoft.imagery",
      "zoom={z}",
      "x={x}",
      "y={y}",
    ];
    const azureMapsUrlTemplate = `${url}?${params.join("&")}`;
    return new XYZ({ url: azureMapsUrlTemplate });
  }
}
```

**Comment by WeatherGod** [comment]
The June 30th deadline for Bing Maps is getting closer. Is there a plan for an AzureMaps to be built in open layers?

**Comment by ahocevar** [comment]
@WeatherGod A pull request would still be welcome.

**Comment by WeatherGod** [comment]
While I'll see what I can do, don't wait on my account. I have very little JS experience and I haven't yet activated an Azure account try out the new map service. Hopefully, a new AzureMap class would mostly look like the old BingMap class, but I'm still not quite sure.

One aspect that might be different is with the subscription key. With BingMaps, it was my understanding that it was tied to the domain(s) it was registered for, so it was pretty common to just include it in your app since anybody else using it anywhere else would fail. But from what I've read so far in the Azure documentation, they do not recommend including the key in the App itself, and that subscription key is tied to a credit card.

I don't know if openlayers has had to deal with this with any other sources. I might also be completely wrong about how I'm reading the documentation and my worries might be overblown.

**Comment by jipexu** [comment]
Hi
I tried to look into this problem. Frankly, the experience provided by the Azure Map portal is pitiful, an unnecessarily and desperately complicated and long interface, all just to get a simple API key... which I still haven't found... portal error... anyway!
All this to say that our OpenLayers examples using BingMaps will soon be obsolete...

I haven't tested this, which seems simple using just an XYZ...?? https://stackoverflow.com/questions/70072448/azure-maps-with-osm-wms-other-layers-in-openlayers 

...finally get a key i will test this ...

**Comment by ahocevar** [comment]
Thanks for your research, @jipexu, and please keep us updated.

**Comment by jipexu** [comment]
I can confirm that with a simple XYZ source, it works fine (I just implemented this in one of my applications...).

https://github.com/azure-samples/azure-maps-openlayers/tree/main/?tab=readme-ov-file#alternative-option-for-openlayers

something like this : 

```
...
new ol.source.XYZ({
               url: `https://atlas.microsoft.com/map/tile?subscription-key=an_api_key_azure&api-version=2.0&tilesetId=one_tileset&zoom={z}&x={x}&y={y}&tileSize=256&language=one_language`,
               attributions: `© ${new Date().getFullYear()} TomTom, Microsoft`
            }
)
...
```

The question now is: should we deimplement our Bing source since an XYZ source is sufficient, or should we modify it?
After that, we can modify or delete our examples depending on the chosen orientation...

The problem is that the API key is necessarily linked to the Azure account's credit card!

**Comment by ahocevar** [comment]
Let's get rid of the BingMaps source, but modify the example to use an ImageTile (the modern alternative to XYZ) source. Since it does not look like we're going to be able to easily maintain an API key, let's let the user enter one, e.g. like in the sentinel-hub example.

**Comment by jipexu** [comment]
ok understand i will try to make something ...

**Comment by jipexu** [comment]
@ahocevar .... i swimming can you have a look here ? https://codesandbox.io/p/devbox/sentinel-hub-forked-rsfkct?workspaceId=ws_64mbKVtKoi1yhXDhy9wpjM

and tell me where i am wrong ... thanks

**Comment by ahocevar** [comment]
What are you trying to do, @jipexu? The code looks correct, but the sandbox does not work for me. Maybe create a draft pull request with what you have, so I can take a look there?

**Comment by jipexu** [comment]
Thanks. I just wanted to test the code in the sandbox (easier for me to see the direct result...) ... but...
Okay, I can try to make a draft pull request...

**Comment by msairam28** [comment]
We are using openLayers to draw BingMaps in our application and since Bing Maps for Enterprise will be retired by end of June 2025, should we migrate away from BingMaps or can we continue using openLayers to draw BingMaps after June 2025.

**Comment by ahocevar** [comment]
Yes @msairam28, you should do that. Either migrate to AzureMaps, or to a different map tiles provider. Both is possible with OpenLayers.

**Comment by M393** [comment]
@msairam28 If you are an enterprise customer for Bing maps, then it should continue to work until [June 30th, 2028](https://learn.microsoft.com/en-us/bingmaps/getting-started/bing-maps-dev-center-help/getting-a-bing-maps-key)
